laser_scan_matcher/src/laser_scan_matcher.cpp

```diff
@@ -770,81 +770,88 @@ void LaserScanMatcher::createCache (const sensor_msgs::LaserScan::ConstPtr& scan
-  pr_ch_x = 0.0;
-  pr_ch_y = 0.0;
-  pr_ch_a = 0.0;
+  tf::Transform pred_last_base_offset = tf::Transform::getIdentity();
 
   // **** use velocity (for example from ab-filter)
   if (use_vel_)
   {
-    pr_ch_x = dt * latest_vel_msg_.linear.x;
-    pr_ch_y = dt * latest_vel_msg_.linear.y;
-    pr_ch_a = dt * latest_vel_msg_.angular.z;
+    double dt = (stamp - last_icp_time_).toSec();
+    // NOTE: this assumes the velocity is in the base frame and that the base
+    //       and laser frames share the same x,y and z axes
+    double pr_ch_x = dt * latest_vel_msg_.linear.x;
+    double pr_ch_y = dt * latest_vel_msg_.linear.y;
+    double pr_ch_a = dt * latest_vel_msg_.angular.z;
 
-    if      (pr_ch_a >= M_PI) pr_ch_a -= 2.0 * M_PI;
-    else if (pr_ch_a < -M_PI) pr_ch_a += 2.0 * M_PI;
+    createTfFromXYTheta(pr_ch_x, pr_ch_y, pr_ch_a, pred_last_base_offset);
   }
 
   // **** use wheel odometry
   if (use_odom_ && received_odom_)
   {
-    pr_ch_x = latest_odom_msg_.pose.pose.position.x -
-              last_used_odom_msg_.pose.pose.position.x;
-
-    pr_ch_y = latest_odom_msg_.pose.pose.position.y -
-              last_used_odom_msg_.pose.pose.position.y;
+    // NOTE: this assumes the odometry is in the base frame
+    tf::Transform latest_odom_pose;
+    tf::poseMsgToTF(latest_odom_msg_.pose.pose, latest_odom_pose);
 
-    pr_ch_a = tf::getYaw(latest_odom_msg_.pose.pose.orientation) -
-              tf::getYaw(last_used_odom_msg_.pose.pose.orientation);
+    pred_last_base_offset = last_used_odom_pose_.inverse() * latest_odom_pose;
 
-    if      (pr_ch_a >= M_PI) pr_ch_a -= 2.0 * M_PI;
-    else if (pr_ch_a < -M_PI) pr_ch_a += 2.0 * M_PI;
-
-    last_used_odom_msg_ = latest_odom_msg_;
+    last_used_odom_pose_ = latest_odom_pose;
   }
 
   // **** use imu
   if (use_imu_ && received_imu_)
   {
-    pr_ch_a = tf::getYaw(latest_imu_msg_.orientation) -
-              tf::getYaw(last_used_imu_msg_.orientation);
+    // NOTE: this assumes the imu is in the base frame
+    tf::Quaternion latest_imu_orientation;
+    tf::quaternionMsgToTF(latest_imu_msg_.orientation, latest_imu_orientation);
 
-    if      (pr_ch_a >= M_PI) pr_ch_a -= 2.0 * M_PI;
-    else if (pr_ch_a < -M_PI) pr_ch_a += 2.0 * M_PI;
+    tf::Quaternion imu_orientation_offset = last_used_imu_orientation_.inverse() * latest_imu_orientation;
+    pred_last_base_offset.setRotation(imu_orientation_offset);
 
-    last_used_imu_msg_ = latest_imu_msg_;
+    last_used_imu_orientation_ = latest_imu_orientation;
   }
+
+  // **** use tf
+  if (use_tf_)
+  {
+    tf::StampedTransform pred_last_base_offset_tf;
+    try
+    {
+      tf_listener_.waitForTransform(base_frame_, last_icp_time_, base_frame_, stamp, fixed_frame_, ros::Duration(tf_timeout_));
+      tf_listener_.lookupTransform (base_frame_, last_icp_time_, base_frame_, stamp, fixed_frame_, pred_last_base_offset_tf);
+      pred_last_base_offset = pred_last_base_offset_tf;
+    }
+    catch (tf::TransformException ex)
```

Comment:
Should this be

     catch (const tf::TransformException& ex)

to do a catch-by-reference instead of a catch-by-value ([C++ guidelines e.14](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#e14-use-purpose-designed-user-defined-types-as-exceptions-not-built-in-types))? Line 785 does this, too, but I'm probably nitpicking since this just goes to logging output.

Reply:
will fix

Reply:
fixed